### DISCUSSION TITLE
unenv: Should we avoid importing from `node:...` from internal modules?

### DISCUSSION THREAD

**Discussion by petebacondarwin** [unjs/unenv#322] [General]
Consider https://github.com/unjs/unenv/blob/e8ea22f9744179c4b79026ba19237bb2c66e2322/src/runtime/node/timers/internal/scheduler.ts#L1-L3

I feel like this code should actually be:

```ts
import { setTimeoutFallbackPromises } from "./set-timeout";
import { setImmediateFallbackPromises } from "./set-immediate";
export class Scheduler implements timers.Scheduler {
  wait(
    delay?: number | undefined,
    options?: Pick<TimerOptions, "signal"> | undefined,
  ) {
    return setTimeoutFallbackPromises(delay, void 0, options);
  }
  yield() {
    return setImmediateFallbackPromises();
  }
}
```

**Comment by petebacondarwin** [comment]
Once we are inside an internal file, I think we should stay there and not pop back out to the polyfilling system.

This is a blocker I hit when trying to implement node.js compat in the Cloudflare Vite Dev server plugin.

**Comment by pi0** [comment]
(moved to unenv discussions)

The main reason for this switch was the starting of hybrid polyfill support. `node:` imports point to the desired polyfill which can be either hybrid or not based on preset configuration.

> This is a blocker I hit when trying to implement node.js compat in the Cloudflare Vite Dev server plugin.

I guess if vite is supporting aliases, it should properly resolve no? We do have it both for Nuxt vite builder behind a flag and Nitro rollup builder.

**Reply by petebacondarwin** [reply]
The problem is that I am using a crazy mixed approach of pre-bundling modules into workerd so that they get used ahead of built-ins (e.g. a module in the workerd bundle called `node:path` will be used instead of the built-in one when imported). But then these unenv entry-points have additional internal imports, which I am handling via the workerd module-fallback mechanism.

For some reason this particular import (`node:timers/promises`) cannot be found from the internally imported file... The simplest approach was to patch unenv, but today I will dig into it a bit more.

---

But to dig into the philosophical position a bit more... the [unenv/src/runtime/node/timers/internal/scheduler.ts](https://github.com/unjs/unenv/blob/e8ea22f9744179c4b79026ba19237bb2c66e2322/src/runtime/node/timers/internal/scheduler.ts) file is only imported from within the `node:timers/promises` package, so one has already committed to being inside the polyfill; importing from `node:times/promises` here is basically a circular dependency, which feels off.

**Reply by pi0** [reply]
I understand we can update this particular case (which self-references).

Generally aliases works. our [workerd tests](https://github.com/unjs/unenv/blob/main/test/workerd/main.mjs) are also implemented with workerd fallabck server and esbuiod preset aliases.
